ENH: stats: Add the Irwin-Hall distribution

nimish opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Depending on your changes, you can skip CI operations and save time and energy: 
http://scipy.github.io/devdocs/dev/contributor/continuous_integration.html#skipping

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Closes #14806
#### What does this implement/fix?
<!--Please explain your changes.-->
Adds the Irwin-Hall distribution i.e. $$\sum^{n}_{k=1}U_k$$ where $U_k \sim U(0, 1)$ and are independent.
#### Additional information
<!--Any additional information you think is important.-->
The Bates distribution is just the IH distribution scaled by $1/n$

nimish committed:
Add the Irwin-Hall distribution

nimish commented:
@ev-br @mdhaber I can't figure out how to have the distribution reject any n that isn't a positive integer, so I didn't add it to the tests.

mdhaber commented:
Did adding this to `_argcheck` not work? Take a look at some of the discrete distributions; some have this constraint.

Note that this would need to be vectorized to support array `n`.

mdhaber commented:
Mostly, but the vectorization needs to work for arbitrary shapes. As written, it only works for 1D arrays.

Also, this needs at least a test against reference values of the CDF or PDF to show that this is indeed the Irwin-Hall distribution. It's OK to let the existing test suite confirm that the distribution is self-consistent (e.g. that the PDF is the derivative of the CDF, etc.).

nimish commented:
> Mostly, but the vectorization needs to work for arbitrary shapes. As written, it only works for 1D arrays.
> 
> 
> 
> Also, this needs at least a test against reference values of the CDF or PDF to show that this is indeed the Irwin-Hall distribution. It's OK to let the existing test suite confirm that the distribution is self-consistent (e.g. that the PDF is the derivative of the CDF, etc.).
> 
> 
> 
> The references also need to be fixed; they are causing issues with the doc build. Check against other distributions. Check your spaces between `..` and the number of the reference; Sphinx is very picky. There should be brackets around the citation numbers, too.
> 
> 
> 
> What is `lol.rst`?

Lol.rst isnt supposed to be there, just a scratch file to test what was throwing the doc parser off.

For tests: I suppose adding a test to confirm the sum of n uniforms has the same pdf and cdf would make sense; I'll add it in.

Equivalenty the mgf being the mgf of the uniform raised to n would also work but I don't see an explicit mgf for that.

mdhaber commented:
> For tests: I suppose adding a test to confirm the sum of n uniforms has the same pdf and cdf would make sense; I'll add it in.

Well, that reminds me that this should override `_rvs` like that. I'm not sure what sort of test you mean. Do you mean comparing the PDF with a convolution of individual uniform PDFs? This would be ok. Or do you mean comparing a normalized histogram (from summing uniformly distributed random samples) against the PDF? The latter would be a pretty weak test.

I would prefer comparison of the output of pdf (at least) against that of [UniformSumDistribution](https://reference.wolfram.com/language/ref/UniformSumDistribution.html).

nimish commented:
> > For tests: I suppose adding a test to confirm the sum of n uniforms has the same pdf and cdf would make sense; I'll add it in.
> 
> Well, that reminds me that this should override `_rvs` like that. I'm not sure what sort of test you mean. Do you mean comparing the PDF with a convolution of individual uniform PDFs? This would be ok. Or do you mean comparing a normalized histogram (from summing uniformly distributed random samples) against the PDF? The latter would be a pretty weak test.
> 
> I would prefer comparison of the output of pdf (at least) against that of [UniformSumDistribution](https://reference.wolfram.com/language/ref/UniformSumDistribution.html).

Added both a K-S GoF (basically a weak histo test....) against a sum of 10 uniforms and the exact values as computed by Wolfram Alpha.

Test suite passes on my machine, at least and I fixed up the sphinx stuff.

nimish commented:
Sigh, apparently not. Let me fix the doc gen

nimish commented:
@mdhaber I only ever get this when I compile docs -- not sure why it's failing on ci

```
Running Sphinx v7.2.6
SciPy (VERSION 1.14.0.dev)
loading pickled environment... done
[autosummary] generating autosummary for: building/blas_lapack.rst, building/compilers_and_options.rst, building/cross_compilation.rst, building/distutils_equivalents.rst, building/index.rst, building/introspecting_a_build.rst, building/redistributable_binaries.rst, building/understanding_meson.rst, dev/api-dev/api-dev-toc.rst, dev/api-dev/array_api.rst, ..., tutorial/stats/probability_distributions.rst, tutorial/stats/quasi_monte_carlo.rst, tutorial/stats/resampling.rst, tutorial/stats/sampling.rst, tutorial/stats/sampling_dau.rst, tutorial/stats/sampling_dgt.rst, tutorial/stats/sampling_hinv.rst, tutorial/stats/sampling_pinv.rst, tutorial/stats/sampling_srou.rst, tutorial/stats/sampling_tdr.rst
[autosummary] generating autosummary for: /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.optimize.LbfgsInvHessProduct.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.bsr_array.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.bsr_matrix.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.csc_array.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.csc_matrix.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.csr_array.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.csr_matrix.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.dok_array.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.dok_matrix.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.lil_array.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.lil_matrix.rst, /Users/nimish/Projects/sdt/scipy/doc/source/reference/generated/scipy.sparse.linalg.LaplacianNd.rst
myst v2.0.0: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=set(), disable_syntax=[], all_links_external=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, fence_as_directive=set(), number_code_blocks=[], title_to_header=False, heading_anchors=0, heading_slug_func=None, html_meta={}, footnote_transition=True, words_per_minute=200, substitutions={}, linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', enable_checkboxes=False, suppress_warnings=[], highlight_code_blocks=True)
myst-nb v1.1.0: NbParserConfig(custom_formats={}, metadata_key='mystnb', cell_metadata_key='mystnb', kernel_rgx_aliases={}, eval_name_regex='^[a-zA-Z_][a-zA-Z0-9_]*$', execution_mode='auto', execution_cache_path='', execution_excludepatterns=['_contents/*.ipynb'], execution_timeout=30, execution_in_temp=False, execution_allow_errors=False, execution_raise_on_error=False, execution_show_tb=False, merge_streams=False, render_plugin='default', remove_code_source=False, remove_code_outputs=False, code_prompt_show='Show code cell {type}', code_prompt_hide='Hide code cell {type}', number_source_lines=False, output_stderr='show', render_text_lexer='myst-ansi', render_error_lexer='ipythontb', render_image_options={}, render_figure_options={}, render_markdown_format='commonmark', output_folder='build', append_css=True, metadata_to_fm=False)
Using jupyter-cache at: /Users/nimish/Projects/sdt/scipy/doc/build/.jupyter_cache
The default value for `navigation_with_keys` will change to `False` in the next release. If you wish to preserve the old behavior for your site, set `navigation_with_keys=True` in the `html_theme_options` dict in your `conf.py` file. Be aware that `navigation_with_keys = True` has negative accessibility implications: https://github.com/pydata/pydata-sphinx-theme/issues/1492
building [mo]: targets for 0 po files that are out of date
writing output...
building [html]: targets for 12 source files that are out of date
updating environment: 0 added, 12 changed, 0 removed
reading sources... [100%] reference/generated/scipy.sparse.linalg.LaplacianNd
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
copying assets... copying static files... done
copying extra files... done
done
writing output... [100%] reference/sparse.linalg
generating indices... genindex done
writing additional pages... search done
copying images... [100%] _static/index_contribute.svg
dumping search index in English (code: en)... done
dumping object inventory... done
[jupyterlite-sphinx] Running JupyterLite build
[jupyterlite-sphinx] JupyterLite build done
build succeeded.

The HTML pages are in build/html.
Skipping build
```

mdhaber commented:
Take a look at the rendered documentation and start by fixing the things that are visually incorrect.

https://output.circle-artifacts.com/output/job/c7a8915e-e41c-43a2-81da-9c9eb77df7b6/artifacts/0/html/reference/generated/scipy.stats.irwinhall.html#scipy.stats.irwinhall

For instance:
![image](https://github.com/scipy/scipy/assets/6570539/417c86db-284e-4283-ad88-61cfd8a48d02)

This is fine for the [extended-summary](https://numpydoc.readthedocs.io/en/latest/format.html#extended-summary), but you need a [one-line summary](https://numpydoc.readthedocs.io/en/latest/format.html#short-summary).

![image](https://github.com/scipy/scipy/assets/6570539/184e77f9-f6ba-402b-9aaa-fd48a7415c55)

As for why the references are not rendering, I mentioned before "Check your spaces between `..` and the number of the reference; Sphinx is very picky. " So `..  [1] HALL PHILIP, "THE` should be `.. [1] HALL PHILIP, "THE`. That might not fix it, but it's addresses a plausible cause of the issue.

fancidev commented:
Out of curiosity, may I ask a few questions related to the distribution:

1. What are the typical magnitude of $n$ in practical use? Is large $n$ practically useful? Does the B-spline formulas work nicely with large $n$?

2. Is there a natural way to extend the definition to positive, real $n$? How would the PDF/CDF formulas look like in such extension? Is such extension practically useful?

I don’t mean these questions need to be addressed by this PR, but I find them theoretically interesting.

Thanks!

nimish commented:
> Out of curiosity, may I ask a few questions related to the distribution:
> 
> 1. What are the typical magnitude of n in practical use? Is large n practically useful? Do the B-spline formulas work nicely with large n?
> 2. Is there a natural way to extend the definition to positive, real n? How would the PDF/CDF formulas look like in such extension? Is such extension practically useful?
> 
> I don’t mean these questions need to be addressed by this PR, but I find them theoretically interesting.
> 
> Thanks!

1. No idea what typical n is, but the B-Spline formula is exact for all n.
2. Sort of, you just add on an extra uniform distribution over 0 .. fractional part of n. I'm not sure of its use.

nimish commented:
@mdhaber rendered docs look clean to me. pdf/cdf/rvs are vectorized manually. I've added the symbolic values computed by Wolfram alpha for n=10 as a check for the distribution.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -9646,6 +9646,53 @@ def test_sf(self):
+        assert_allclose(ih_cdf, 0.5)
+
+    ih10 = stats.irwinhall(10)
+    # from Wolfram Alpha: "values of pdf of UniformSumDistribution[10] for x=0 to x=10"
+    # symmetric about n/2 = 5
+    vals = [0, 1/362880, 251/181440, 913/22680, 44117/181440]
+    vals += [15619/36288] + vals[::-1]
```
Just confirming:
<img width="758" alt="image" src="https://github.com/scipy/scipy/assets/6570539/a7e46c34-5991-4c8a-9064-f3fee06b86f2">
👍

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -9646,6 +9646,53 @@ def test_sf(self):
+    assert_allclose(pdf_ih2, expected) # should be a triangle
+
+    # PDF is symmetric so CDF should be 0.5 at n/2
+    for n in range(2, 10):
+        ih = stats.irwinhall(n)
+        ih_cdf = ih.cdf(n/2)
+        assert_allclose(ih_cdf, 0.5)
```
```suggestion
    n = np.arange(1, 10)
    ih = stats.irwinhall(n)
    ih_cdf = ih.cdf(n/2)
    assert_allclose(ih_cdf, 0.5)
```

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -9646,6 +9646,53 @@ def test_sf(self):
+    pts = np.linspace(0, 1, 100)
+    pdf_unif = unif.pdf(pts)
+    pdf_ih1 = ih1.pdf(pts)
+    assert_allclose(pdf_ih1, pdf_unif) # should be identical to uniform
```
```suggestion
    assert_allclose(pdf_ih1, pdf_unif)  # should be identical to uniform
```
Minor, and I guess the linter doesn't enforce it, but would be preferred per [PEP8](https://peps.python.org/pep-0008/#inline-comments).

mdhaber replied:
I only added the comment in one place, but it is applicable elsewhere.

nimish replied:
Hm I'm going off the linter; does it need updating?

mdhaber replied:
The rule was enforced by our old linter, but it is in a beta state in Ruff:
https://docs.astral.sh/ruff/rules/too-few-spaces-before-inline-comment/

Even when it becomes officially supported, we might not add it - there are probably a lot of exceptions in the codebase, and we don't want people to have to fix unrelated lines whenever they touch a file. But it's easy enough to enforce in new code on a best-effort basis.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -9646,6 +9646,53 @@ def test_sf(self):
+    assert_allclose(cdf_ih1, cdf_unif) # should be identical to uniform
+
+    ih2 = stats.irwinhall(2)
+    pdf_ih2 = ih2.pdf(1)
+    assert_allclose(pdf_ih2, 1) # should be 1 at x=1
+
```
```suggestion
```
I think that's is pretty much covered by the next part.

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -9646,6 +9646,53 @@ def test_sf(self):
+    pdf_ih2 = ih2.pdf(1)
+    assert_allclose(pdf_ih2, 1) # should be 1 at x=1
+
+    npts = 100
+    pts = np.linspace(0, 2, npts)
+    expected = np.linspace(0, 2, npts)
+    expected[npts//2:] = 2 - expected[npts//2:]
```
```suggestion
    npts = 101
    pts = np.linspace(0, 2, npts)
    expected = np.linspace(0, 2, npts)
    expected[(npts+1)//2:] = 2 - expected[(npts+1)//2:]
```
If you want to test the midpoint, I think this is right.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9033,6 +9034,101 @@ def _munp(self, n, b):
+    with knots evenly spaced from :math:`1` to :math:`n`.
+
+    The Bates distribution is simply the Irwin-Hall distribution scaled 
+    by :math:`1/n`: ``bates = irwinhall(n, name="bates", scale=1/n)``
```
Does this actually work? I don't know if you can freeze a distribution with a new name.
```python3
from scipy.stats import irwinhall
n = 10
bates = irwinhall(n, name="bates", scale=1/n)
# TypeError: _parse_args() got an unexpected keyword argument 'name'
```

nimish replied:
Hm. I didn't see any issues. How else do you make parametrically specialized versions of general distributions?

mdhaber replied:
> I didn't see any issues.

OK, but please try executing the code to see if you get the same error.

> How else do you make parametrically specialized versions of general distributions?

I'm not immediately seeing a way to do this such that the result object `bates` behaves exactly like a regular distribution object. It is easy to implement a `bates` such that the object is a "frozen" distribution, which is basically what you have.
```python3
bates = irwinhall(n, scale=1/n)
```
There is probably some way to do it, but it is probably not worth figuring out right now. It is worth including this in the notes of the documentation, though.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9033,6 +9034,101 @@ def _munp(self, n, b):
+    @staticmethod
+    def _cardbspl(n):
+        t = np.arange(n+1) 
+        return BSpline.basis_element(t)
```
Same here - so that I can confirm, please add a reference and page number.

mdhaber replied:
Please leave these unresolved; I'll resolve them when I've confirmed that they have been addressed.

How has this been addressed? I would appreciate a comment right here in the code pointing to the reference number. As a reviewer, I need to do due diligence and check out the source of all this stuff. We all have different backgrounds, and I am not intimitely familiar with this distribution or B-splines, so the connection is not immediately obvious. I can also look it up, but this takes more time, and of course there will be others who inspect this code in the future who would need to do the same. Let's make it easier for everyone - except you, sorry : )

mdhaber replied:
@ev-br @steppi I could dive into B-splines and such, but I wonder if you'd be willing to review this part - that the PDF of the Irwin-Hall distribution PDF is as simple as this, no gotchas that you see?

nimish replied:
@mdhaber take a look at the doc references 4 and 5: from https://www.chebfun.org/examples/approx/BSplineConv.html (by @trefethen and chebfun/chebfun )


>B-splines were introduced by Schoenberg and became a standard method for numerical computation following the work of de Boor [1] and Cox [2] in 1972. In general there is no need for the breakpoints to be uniformly spaced. With uniformly spaced points as in our example, however, the shape approaches a Gaussian as n→∞. One could derive this fact by noting that in Fourier space, each convolution is a multiplication by the Fourier transform of B0, namely a sinc function. This is the same computation that arises in the proof of the Central Limit Theorem, and in fact, the B-splines have a probabilistic interpretation: if B0 represents a random variable with uniform distribution on the interval [−1/2,1/2] then B1 is the distribution for the sum of two independent samples from B0, B2 is the distribution for the sum of three samples from B0, and so on.

nimish replied:
Another derivation comes from the cumulants of the uniform. Since the IH is just a sum of `n` standard uniforms, you take `n` times the uniform cumulant.

The nth cumulant of U(0,1) is $\frac{B_n}{n}$ where $B_n$ is nth [Bernoulli number](https://en.wikipedia.org/wiki/Bernoulli_number) so the nth cumulant of an IH(m) distribution is $$\frac{mB_n}{n}$$

That immediately implies the odd order cumulants are 0 for $n \geq 3$. We could use `scipy.special.bernoulli` for the higher order ones. 

The formulas for moments in terms of cumulants quickly get very complicated, but it could be done with `sympy`

mdhaber replied:
Yeah, that last part is good. Their BSplines are defined on `[−1/2,1/2]` and ours are on `[0, 1]`, I guess?

nimish replied:
Yes but it doesn't matter due to translation invariance (fortunately) -- basically the equivalent of setting `loc=1/2`

steppi replied:
@mdhaber, this looks right to me. I'll sign off on this. The reference @nimish provided looks good too.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9033,6 +9034,101 @@ def _munp(self, n, b):
+
+    References
+    ----------
+    .. [1] HALL PHILIP, "THE DISTRIBUTION OF MEANS FOR SAMPLES OF SIZE N DRAWN 
+            FROM A POPULATION IN WHICH THE VARIATE TAKES VALUES BETWEEN 0 AND 1, 
+            ALL SUCH VALUES BEING EQUALLY PROBABLE", 
```
```suggestion
    .. [1] P. Hall, "The distribution of means for samples of size N drawn
            from a population in which the variate takes values between 0 and 1,
            all such values being equally probable",
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9033,6 +9034,101 @@ def _munp(self, n, b):
+            ALL SUCH VALUES BEING EQUALLY PROBABLE", 
+            Biometrika, Volume 19, Issue 3-4, December 1927, Pages 240-244, 
+            https://doi.org/10.1093/biomet/19.3-4.240
+    .. [2] J. O. IRWIN, "ON THE FREQUENCY DISTRIBUTION OF THE MEANS OF SAMPLES 
+            FROM A POPULATION HAVING ANY LAW OF FREQUENCY WITH FINITE MOMENTS, 
+            WITH SPECIAL REFERENCE TO PEARSON'S TYPE II", 
```
```suggestion
    .. [2] J. O. IRWIN, "On the frequency distribution of the means of samples
            from a population having any law of frequency with finite moments,
            with special reference to Pearson's Type II,
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9033,6 +9034,101 @@ def _munp(self, n, b):
+            FROM A POPULATION HAVING ANY LAW OF FREQUENCY WITH FINITE MOMENTS, 
+            WITH SPECIAL REFERENCE TO PEARSON'S TYPE II", 
+            Biometrika, Volume 19, Issue 3-4, December 1927, Pages 225-239, 
+            https://doi.org/10.1093/biomet/19.3-4.225
```
```suggestion
            :doi:`0.1093/biomet/19.3-4.225`
```

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9033,6 +9034,101 @@ def _munp(self, n, b):
+            FROM A POPULATION IN WHICH THE VARIATE TAKES VALUES BETWEEN 0 AND 1, 
+            ALL SUCH VALUES BEING EQUALLY PROBABLE", 
+            Biometrika, Volume 19, Issue 3-4, December 1927, Pages 240-244, 
+            https://doi.org/10.1093/biomet/19.3-4.240
```
```suggestion
            :doi:`10.1093/biomet/19.3-4.240`
```

mdhaber reviewed (commented):
Looks really close once tests pass. The main thing is to add citations as code comments where there is a formula. I'll run the tests, but I expect `fit` to fail; we'll have to decide what to do about that.

nimish commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9033,6 +9034,101 @@ def _munp(self, n, b):
+    def _rvs(self, n, size=None, random_state=None, *args):
+        @_vectorize_rvs_over_shapes
+        def _rvs1(n, size=None, random_state=None):
+            if not self._argcheck(n):
+                raise ValueError("n must be a positive integer")
```
Oddly, no -- needed to actually enforce the integrality so that the size trick and sum works

mdhaber replied:
I don't understand. I removed this code, and then all three of these:
```python3
irwinhall(-1).rvs()
irwinhall(1.5).rvs()
irwinhall.rvs(-1, size=None)
irwinhall.rvs(1.5, size=None)
```
produce the desired, generic error.
```
ValueError: Domain error in arguments. The `scale` parameter must be positive for all distributions, and many distributions have restrictions on shape parameters. Please see the `scipy.stats.irwinhall` documentation for details.
```
So why is this needed here in `rvs`, and especially why in the loop itself?

nimish replied:
I'll grab the error I get

nimish replied:
Eventually just removed it.

mdhaber replied:
It produces an appropriate error, right? (If it does on my computer but not yours, that's surprising!)

nimish commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9033,6 +9034,101 @@ def _munp(self, n, b):
+        return _rvs1(n, size=size, random_state=random_state)
+    
+    def _stats(self, n):
+        # mgf = ((exp(t) - 1)/t)**n
+        # m'th derivative follows from the generalized Leibniz rule
+
```
To be honest it derives trivially from the definition as the sum of n independent uniform rvs which implies that the MGF is just the uniform one to the nth power

mdhaber replied:
We all have different backgrounds, and so what is trivial to some (https://github.com/scipy/scipy/pull/20481#discussion_r1582587592) may not be clear to all. Thanks for extending the comment, though. I'll go ahead and add the direct reference for those who would prefer to avoid the derivation.

nimish commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9033,6 +9034,101 @@ def _munp(self, n, b):
+
+    Conveniently, the pdf and cdf are the :math:`n`-fold convolution of 
+    the ones for the standard uniform distribution,
+    which are well known to be just cardinal B-splines of degree :math:`n-1` 
+    with knots evenly spaced from :math:`1` to :math:`n`.
```
Yeah it's from the Wikipedia page but also just from the convolution theorem. Probably requires familiarity with spline definitions. I'll see if I can add a better citation

nimish committed:
revert since the other way fails tests

nimish committed:
skip fit for ih

nimish committed:
redundant

mdhaber commented:
BTW, no more force-pushes please. We will squash merge this at the end, and I need to see what has changed in the history.

Also, you can include `[skip ci]` in the body of commit messages if you don't need to re-run CI.

nimish commented:
> BTW, no more force-pushes please. We will squash merge this at the end, and I need to see what has changed in the history.
> 
> Also, you can include [`[skip ci]`](https://docs.scipy.org/doc/scipy/dev/contributor/continuous_integration.html#skipping) in the body of commit messages if you don't need to re-run CI. Please do that to conserve resources when appropriate.

@mdhaber 
Sure, will do both.

Though I'd add config to only ever run a single CI run at a time for PRs and/or give the PR author the ability to cancel the run if done accidentally. Personally I'd also set it to just manual runs if it's that much of an issue/run only the shallowest suite automatically.

Anything left to do before merging?

mdhaber commented:
> Though I'd add config to only ever run a single CI run at a time for PRs and/or give the PR author the ability to cancel the run if done accidentally.

Current most CI is not running automatically on this PR, I think because this is your first. It's just a good habit to get into for when it does run by default.

nimish committed:
[noci] lint and docs

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9033,6 +9034,101 @@ def _munp(self, n, b):
+        return [_ShapeInfo("n", True, (1, np.inf), (True, False))]
+    
+    def _munp(self, n, t):
+        return (sc.expm1(t)/t)**n
```
OK, I verified that the MGF of the uniform distribution on the unit interval is $\frac{e^t - 1}{t}$ and that $M_{X+Y} = M_x \cdot M_y$, so $M_{nX} = M_x^n$. 

However, `_munp` is not the moment generating function; it is the raw moment about the origin. The first argument, `n`, is the order of the moment, not the shape parameter `n` of the distribution. Is that what you were expecting?

In any case, this fails the generic moment test:

```
_________________________________ test_moments[irwinhall-arg50-True-True-True-False] __________________________________

distname = 'irwinhall', arg = (10,), normalization_ok = True, higher_ok = True, moment_ok = True, is_xfailing = False

    @pytest.mark.slow
    @pytest.mark.parametrize('distname,arg,normalization_ok,higher_ok,moment_ok,'
                             'is_xfailing',
                             cases_test_moments())
    def test_moments(distname, arg, normalization_ok, higher_ok, moment_ok,
                     is_xfailing):
        try:
            distfn = getattr(stats, distname)
        except TypeError:
            distfn = distname
            distname = 'rv_histogram_instance'

        with npt.suppress_warnings() as sup:
            sup.filter(IntegrationWarning,
                       "The integral is probably divergent, or slowly convergent.")
            sup.filter(IntegrationWarning,
                       "The maximum number of subdivisions.")
            sup.filter(IntegrationWarning,
                       "The algorithm does not converge.")

            if is_xfailing:
                sup.filter(IntegrationWarning)

            m, v, s, k = distfn.stats(*arg, moments='mvsk')

            with np.errstate(all="ignore"):
                if normalization_ok:
                    check_normalization(distfn, arg, distname)

                if higher_ok:
                    check_mean_expect(distfn, arg, m, distname)
                    check_skew_expect(distfn, arg, m, v, s, distname)
                    check_var_expect(distfn, arg, m, v, distname)
                    check_kurt_expect(distfn, arg, m, v, k, distname)
>                   check_munp_expect(distfn, arg, distname)

arg        = (10,)
distfn     = <scipy.stats._continuous_distns.irwinhall_gen object at 0x00000207EB40AED0>
distname   = 'irwinhall'
higher_ok  = True
is_xfailing = False
k          = -0.12
m          = 5.0
moment_ok  = True
normalization_ok = True
s          = 0.0
sup        = <numpy.testing._private.utils.suppress_warnings object at 0x00000207ECED4B10>
v          = 0.8333333333333334

scipy\stats\tests\test_continuous_basic.py:348:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
scipy\stats\tests\common_tests.py:102: in check_munp_expect
    assert_allclose(res, ref, atol=1e-10, rtol=1e-4,
        args       = (10,)
        dist       = <scipy.stats._continuous_distns.irwinhall_gen object at 0x00000207EB40AED0>
        msg        = 'irwinhall'
        ref        = 4216.6666666727815
        res        = 5.183528709112784e+16
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

args = (<function assert_allclose.<locals>.compare at 0x00000207EC9239C0>, array(5.18352871e+16), array(4216.66666667))
kwds = {'equal_nan': True, 'err_msg': 'irwinhall - higher moment / _munp', 'header': 'Not equal to tolerance rtol=0.0001, atol=1e-10', 'verbose': True}

    @wraps(func)
    def inner(*args, **kwds):
        with self._recreate_cm():
>           return func(*args, **kwds)
E           AssertionError:
E           Not equal to tolerance rtol=0.0001, atol=1e-10
E           irwinhall - higher moment / _munp
E           Mismatched elements: 1 / 1 (100%)
E           Max absolute difference: 5.18352871e+16
E           Max relative difference: 1.22929535e+13
E            x: array(5.183529e+16)
E            y: array(4216.666667)

args       = (<function assert_allclose.<locals>.compare at 0x00000207EC9239C0>, array(5.18352871e+16), array(4216.66666667))
func       = <function assert_array_compare at 0x000002076FAA7D80>
kwds       = {'equal_nan': True, 'err_msg': 'irwinhall - higher moment / _munp', 'header': 'Not equal to tolerance rtol=0.0001, atol=1e-10', 'verbose': True}
self       = <contextlib._GeneratorContextManager object at 0x000002076F9B1510>

..\miniforge3\envs\scipy-dev\Lib\contextlib.py:81: AssertionError
=============================================== short test summary info ===============================================
FAILED scipy/stats/tests/test_continuous_basic.py::test_moments[irwinhall-arg50-True-True-True-False] - AssertionError:
================================ 1 failed, 108 passed, 10 skipped, 2 xfailed in 26.85s ================================
```

When I remove this function override, the test passes because it falls back to the default `_munp` implementation. So the simplest fix is just to remove this unless you would like to work out the formula.

mdhaber replied:
Let's leave it out for now, and if you want to submit it as a follow-up PR, we could consider it.

nimish replied:
Well, ended up doing the derivation by hand and it turns out to be surprisingly simple in the end--let me see if I can add it in. Otherwise I'll just defer to the superclass.

@mdhaber we definitely don't do anything with cumulants, right? If not we probably should since they tend to work for cases where moments don't and converting between the two is tedious but doable.

mdhaber replied:
Feel free to make that suggestion in gh-15928. It is already on my mind to add support for distribution L-moments (you can see that I've drafted support for the distribution of order statistics), and I've had a PR open since the fall for a function to compute sample l-moments (gh-19475). Both of these are toward the goal of adding method of L-moments for fitting.

mdhaber replied:
> Well, ended up doing the derivation by hand and it turns out to be surprisingly simple in the end

You seemed interested in getting this PR merged sooner rather than later. Adding these will slow it down because it would require a maintainer not only to review the new code but also follow your derivations (which probably means deriving them oneself).

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -9646,6 +9646,53 @@ def test_sf(self):
+    vals += [15619/36288] + vals[::-1]
+    assert_allclose(ih10.pdf(np.arange(11)), vals)
+
+    # check that the definition of the Irwin-Hall distribution is correct
+    nsamp = 1000
+    uu = unif.rvs(size=nsamp, random_state=1234)
+    sum10s = np.bincount(np.arange(len(uu))//10, uu)
+    res = stats.kstest(sum10s, ih10.cdf)
+    assert res.pvalue > 0.1
+
+
+
+
```
I see. So even if they currently execute pretty much the same test, the _intent_ of this one is to be a weak test of the CDF implementation, whereas the intent of `check_distribution_rvs` is to be a test of consistency between the RVS and CDF.

In any case, you have a much stronger test of the CDF above (i.e. comparing actual values against Wolfram Alpha). So unless the previous test fails, I can only see this test producing false positives. You don't need to remove it; I can do that too.

nimish replied:
The idea was to check the distributional relation between IH(N) and summing N iid uniforms. As far as I can tell the other checks test the random variable generation against its own cdf implementation vs the cdf of a different distribution.

Practically they should always pass since we do the sum of uniforms way to make IH samples, but suppose someone changed that to use the inverse of the cdf or some other generation algorithm (ziggurat?) instead. 

We'd have to make sure that summing N uniforms and taking a single IH(N) sample are distributionally identical or close to it.

I guess it's like testing that chi-square is really the distribution of a squared normal r.v. This should hold regardless of how normal and chi-square are implemented.

nimish replied:
Well, to [quote Knuth](https://staff.fnwi.uva.nl/p.vanemdeboas/knuthnote.pdf) on this: 
>Beware of bugs in the above code; I have only proved it correct, not tried it.

I'll remove it for now

mdhaber commented on scipy/stats/tests/test_continuous_basic.py:
```diff
@@ -76,7 +76,7 @@
                'gausshyper', 'gengamma', 'genhyperbolic',
                'geninvgauss', 'gennorm', 'genpareto',
-               'halfcauchy', 'halfgennorm', 'invgamma', 'jf_skew_t',
+               'halfcauchy', 'halfgennorm', 'invgamma', 'irwinhall', 'jf_skew_t',
```
> Yes, it's not super clear to me why this fails.

If "this" means a complex derivative test, it makes sense to me because BSplines don't support complex inputs.

As for why I would not expect the `fit` method to work, it's because it relies on a generic optimization procedure that does not consider (and can't really handle) the integrality constraint. If `scipy.stats.fit` is failing, I'd need to take a closer look. 

I'd like to get other opinions about the `fit` method being unable to fit the shape parameter. The `fit` method is not guaranteed to work for all possible distributions, data, and guesses, but in this case we are essentially guaranteed that it will fail and will just produce garbage. It would probably be better for users if we raise an error unless the shape parameter is fixed. (It can probably fit location and scale, though.)

nimish replied:
I've added it to the skip/xfail lists for now

mdhaber commented:
> Anything left to do before merging?

Oops, yes - thanks for your patience!

All new features need an announcement at https://discuss.scientific-python.org/c/contributor/scipy/32. We used to do this on a mailing list; here is an [example](https://mail.python.org/archives/list/scipy-dev@python.org/message/WLBUERFYC2GOPCFHABRXATNDYJFHLGEN/). One thing I imagine that people might want to comment on is the name - `irwinhall` vs, say, `uniformsum`, so that might be something to ask about in the email.

fancidev commented:
Overriding `fit` that checks `n` is fixed sounds appropriate. That is also why I asked if the distribution can be “easily” extended to allow real `n`.

`UniformSum` sounds good to me, but I’ve  never used or heard of the distribution. Not sure if `IrwinHall` or `Bates` would be more familiar to people who actually use them.

nimish commented:
> Overriding `fit` that checks `n` is fixed sounds appropriate. That is also why I asked if the distribution can be “easily” extended to allow real `n`.
> 
> 
> 
> `UniformSum` sounds intuitive to me (and it’s the name picked by [Mathematica](https://reference.wolfram.com/language/ref/UniformSumDistribution.html)), but I’ve  never used or heard of the distribution. Not sure if `IrwinHall` or `Bates` would be more familiar to people who actually use them.

Doing the obvious extension to non integral 'n' by adding an extra weighted uniform breaks a lot of the nice formulas. 

I could see adding a UniformSum that did accept non integer N and keeping the IH as the integral only version 

IME neither seem to be used that often over just treating the sum as a sum.

mdhaber commented:
Re: implementing for non-integral `n` - please let's not attempt that for now.

nimish committed:
address fitting and moments

nimish committed:
[noci] delete KS test

Co-authored-by: Matt Haberland <mhaberla@calpoly.edu>

nimish commented:
1. removed argcheck
2. I've added a commented-out but correct (checked against W|A) implementation of the moments calculated from the cumulants that really should live in the generic `rv_continuous` implementation. It's debatable whether it's better to do that over the generic computation. Some other distributions have very easily calculated cumulants and difficult moments like this one.
3. Removed the bates comment
4. `fit` will throw an error and has been added to the xfail/skip lists

mdhaber committed:
Merge branch 'main' into irwinhall

mdhaber commented:
Great. Please remember to send that announcement. 
https://github.com/scipy/scipy/pull/20481/#issuecomment-2089299532

nimish commented:
> > Anything left to do before merging?
> 
> Oops, yes - thanks for your patience!
> 
> All new features need an announcement at [discuss.scientific-python.org/c/contributor/scipy/32](https://discuss.scientific-python.org/c/contributor/scipy/32). We used to do this on a mailing list; here is an [example](https://mail.python.org/archives/list/scipy-dev@python.org/message/WLBUERFYC2GOPCFHABRXATNDYJFHLGEN/). One thing I imagine that people might want to comment on is the name - `irwinhall` vs, say, `uniformsum`, so that might be something to ask about in the post. Another thing to get other opinions about is what to do about the `fit` method.
> 
> Besides that, I un-resolved all the comments above and re-resolved the ones that are resolved. We just need to address the ones that are still unresolved:
> 
> * [x]  We need a decision [here](https://github.com/scipy/scipy/pull/20481#discussion_r1586920076) about what to do with the `fit` method, given that it's essentially _guaranteed_ to fail unless the shape parameter is fixed. I'd suggest overriding `fit`, raising a `NotImplementedError` unless the shape is fixed, and falling back to `super().fit` otherwise. (_Update: always raising `NotImplementedError` is fine, too._)
> * [x]  The [comment ](https://github.com/scipy/scipy/pull/20481#discussion_r1586893167) about the manual `_argcheck` within the loop of the `rvs` method is still open.
> * [x]  According to [this comment](https://github.com/scipy/scipy/pull/20481#discussion_r1586875855), the `_munp` implementation needs to be corrected or removed. (_And I'll need to remove the commented-out code before merging._)
> * [ ]  I can add the reference for the `stats` method that I [requested](https://github.com/scipy/scipy/pull/20481#discussion_r1582606542), remove the [redundant test](https://github.com/scipy/scipy/pull/20481#discussion_r1586898239), and fix the [PEP8 issue](https://github.com/scipy/scipy/pull/20481#discussion_r1586900191); don't worry about those.
> * [x]  According to [this](https://github.com/scipy/scipy/pull/20481#discussion_r1586907868), we need to adjust the comment about how to create a Bates distribution. (_I'll add an improved comment before merging._)
> * [x]  It would help for someone more familiar with B-splines to [review the PDF implementation](https://github.com/scipy/scipy/pull/20481#discussion_r1586925239). (I absolutely believe it; it's just not immediately obvious to me, and there are many things to do!) (_Update: Thanks @steppi!_)
> * [ ]  We'll need to make sure CI is green before merging. I'm running it so we can see if there's anything else that fails besides the moment check. (_Update: I'll fix the lint issues before merging._)
> * [ ]  Approval from discussion forum.
> 
> Thanks @nimish!

Re the moments/cumulants code: I'll remove it.

nimish committed:
[noci] delete commented out code

mdhaber commented:
@nimish Great, thanks! Re:

> Adding that will be a separate feature because scipy needs more cumulant infrastructure (which I’ll look at adding)

I'd recommend taking a look at gh-15928 and commenting there before writing code.

rlucas7 commented:
@nimish thanks for working on this distribution. I'm not sure what the state of the distributions infrastructure is @mdhaber or @steppi please chime in if I'm derailing anything here. 

@nimish for the general moments of this distribution you should be able to use the formula I noted [here](https://github.com/scipy/scipy/issues/17890#issue-1563508767). That formula should work for all non-central non-negative integer moments as far as I can tell. 

The binomial coefficients are available from `scipy.special` (in approximate and exact form), as are the Stirling numbers of the second kind, for the latter they were added here #18103 

I'll drop a note with a backlink here in the discuss forum.

mdhaber commented:
Not derailing, since the new infrastructure is a different issue. If the formula is that simple in terms of the Stirling numbers, it would be OK to include them in this PR, accompanied by the citation information.

nimish commented:
> @nimish thanks for working on this distribution. I'm not sure what the state of the distributions infrastructure is @mdhaber or @steppi please chime in if I'm derailing anything here. 
> 
> 
> 
> @nimish for the general moments of this distribution you should be able to use the formula I noted [here](https://github.com/scipy/scipy/issues/17890#issue-1563508767). That formula should work for all non-central non-negative integer moments as far as I can tell. 
> 
> 
> 
> The binomial coefficients are available from `scipy.special` (in approximate and exact form), as are the Stirling numbers of the second kind, for the latter they were added here #18103 
> 
> 
> 
> I'll drop a note with a backlink here in the discuss forum. 

Thanks. I have the code and derivation of a general cumulant to moment formula but left it out for now

rlucas7 commented:
> It looks like the stirling number formula for the moments lines up with the one from the cumulants so I'll stick the simpler one from @rlucas7 in, thanks!

Sounds good @nimish I agree with @mdhaber 's comment here: 

> I'd do some informal tests of the useful domain of the formula. When do the terms begin to overflow? And in what cases do we get low precision (when is the result close to 1)? Does the generic implementation do better in these cases?

I'd guess that at some higher `n` (number of uniforms summed) and `k` raw moment power) the Stirling numbers of the second kind and the Binomial coefficients both are large and the division into floating point should be checked. 

I'm guessing the values for this scenario are larger than where I'd think to use but in any case we usually try to document where the numerics are less accurate. (you can see examples of this in the pr for stirling2 where I wrote the approximate branch of the code). 

Note: for the moments here I'd recommend to use the `exact=True` branch of binomial and stirling2.

nimish commented:
Makes sense. Worst case scenario i fall back to logs or other tricks

mdhaber committed:
MAINT: stats.irwinhall: adjustments per review

mdhaber commented:
Most recent commit addresses my comments and should fix the lint issues. I'll also run the xslow tests locally to make sure all the right ones are xslowed or skipped.

WarrenWeckesser commented:
This is not a full review of the pull request, but I can see some changes that should be made.

* An explicit `_sf` method should be created, otherwise a call such as `irwinhall(10).sf(9.9)` will return `0.0` instead of the correct `2.755731922398591e-17`.  It can be implemented to take advantage of the symmetry of the distribution.
* All uses of `assert_allclose` should have an explicit `rtol` given, and generally it should be much smaller than the default `1e-7`. `rtol=1e-14` is a typical value to start with (loosen as needed, but hopefully we don't need to approach the default value).  @nimish, this suggestion is as much for all the other SciPy developers as it is for you--this suggestion should probably be in the "missing bits" developer docs, if it is not already written down somewhere.  I know it has come up in a review or two over the last few years.
* Currently there is a single test function `test_irwinhall()` that tests *everything*.  This should be split into several shorter test functions, where each test is for a specific method (e.g. `test_irwinhall_pdf()`, `test_irwinhall_cdf()`, etc).  Modularizing the tests like that makes it much easer to read the reports generated by `pytest` when a test fails.  It also allows all the tests to be run, instead of failing at the first failed assertion.  For most (all?) of the distributiobns, we define a class to contain the tests (e.g, `TestBinom`, `TestArcsine`, etc; see the first several thousand lines of `test_distributions.py`).  We don't follow that pattern exclusively (I think I've add several test functions over the years that could have been a method of an existing class), but for a new distribution, we should follow the old pattern--for consistency and for the organizational convenience--and create a class (e.g. `class TestIrwinHall`) with the individual tests as methods.
* A test of the CDF that checks the values against values from Wolfram Alpha should be added.

mdhaber commented:
@WarrenWeckesser are you willing to take over review here? I'd suggest participating in gh-17807 if want to impose these requirements on all distributions. I don't necessarily agree that these things should be in the distribution specific test, and I can write a full response later.

nimish commented:
@WarrenWeckesser good point on the numeric cancellation, that's a genuine bug I'll look at fixing. But as @mdhaber said catastrophic cancellation at the tails of a symmetric distribution is an infra-wide problem that likely needs regression testing and should be enforced at a higher level.

* `rtol`: this is debatable, and really a project wide policy thing since the semantics of `isclose` are subtle. I'll keep this at the default unless there's a good reason not to.
* I can split up the tests into individual functions. We don't actually use much of the class functionality in pytest (again a higher level stats infra problem to tackle) so I didn't see a need.

nimish committed:
address pr comments

nimish committed:
add sf

nimish committed:
add moments and precision tests

We should be within 10 ulps of the exact rational division,
but this boils down to the precision of bspline evaluation
more than anything

nimish commented:
@WarrenWeckesser / @mdhaber 

Long story short, the best we can do here is to be equal within 10 ulps (most functions are better). Unfortunately, since the exact value of `cdf(99/100) = 36287999999999999/36288000000000000` is within an ulp of 1, even wolframalpha will round it to `1.0` unless specifically asked to evaluate it using rationals.

I think we're reasonably close and even then it's tough to justify any closer without heuristics that might be better placed in global infrastructure.

mdhaber committed:
Merge remote-tracking branch 'upstream/main' into irwinhall

mdhaber reviewed (approved):
Recent commits LGTM. I'll merge main to see if the Circle failure was a fluke and run the other tests.

mdhaber commented on scipy/stats/_continuous_distns.py:
```diff
@@ -9036,6 +9037,132 @@ def _munp(self, n, b):
+        # see https://link.springer.com/content/pdf/10.1007/s10959-020-01050-9.pdf
+        # page 640, with m=n, j=n+order
+        def vmunp(order, n):
+            return sc.stirling2(n+order, n, exact=True) / sc.comb(n+order, n, exact=True)
+
```
```suggestion
            return (sc.stirling2(n+order, n, exact=True) 
                    / sc.comb(n+order, n, exact=True))

```

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -9669,6 +9671,99 @@ def test_sf(self):
+
+    def test_pdf_ih1_unif(self):
+        # IH(1) PDF is by definition U(0,1)
+        # we should be too, but differences in floating point eval order happen
+        # it's unclear if we can get down to the single ulp for doubles unless quads are used
+        # we're within 6-10 ulps otherwise (across sf/cdf/pdf) which is pretty good
```
```suggestion
        # we should be too, but differences in floating point eval order happen
        # it's unclear if we can get down to the single ulp for doubles unless
        # quads are used we're within 6-10 ulps otherwise (across sf/cdf/pdf) 
        # which is pretty good
```

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -9669,6 +9671,99 @@ def test_sf(self):
+        assert_array_max_ulp(self.ih10.cdf(np.arange(11)), vals, maxulp=10)
+
+        assert_array_max_ulp(self.ih10.cdf(1/10), 1/36288000000000000, maxulp=10)
+        assert_array_max_ulp(self.ih10.cdf(99/10), 36287999999999999/36288000000000000, maxulp=10)
```
```suggestion
        ref = 36287999999999999/36288000000000000
        assert_array_max_ulp(self.ih10.cdf(99/10), ref, maxulp=10)
```

mdhaber commented on scipy/stats/tests/test_distributions.py:
```diff
@@ -9669,6 +9671,99 @@ def test_sf(self):
+        # from Wolfram Alpha "SurvivalFunction[UniformSumDistribution[10],x] at x=1/10"
+        # and symmetry about n/2 = 5
+        # W|A returns 1 for CDF @ x=9.9
+        assert_array_max_ulp(self.ih10.sf(1/10), 36287999999999999/36288000000000000, maxulp=10)
```
```suggestion
        ref = 36287999999999999/36288000000000000
        assert_array_max_ulp(self.ih10.sf(1/10), ref, maxulp=10)
```

mdhaber reviewed (commented):
To address lint failure

mdhaber committed:
Apply suggestions from code review

[skip ci]

mdhaber committed:
BENCH: stats.irwinhall: skip fit benchmark

[skip cirrus]

mdhaber commented:
CI is green, and no more feedback, so in it goes. Thanks @nimish et al!

lucascolley commented:
if anyone would like to write a release note for this, please do!

mdhaber commented:
I always go through and write notes as we approach release rather than as PRS are merged, and I think I saw that only maintainers can edit the wiki now?

nimish commented:
> I always go through and write notes as we approach release rather than as PRS are merged, and I think I saw that only maintainers can edit the wiki now?

I can do this if you've got any good examples

lucascolley commented:
> I can do this if you've got any good examples

Just something to appear under '`scipy.stats` improvements' in the release notes, e.g. https://github.com/scipy/scipy/releases/tag/v1.13.0.

> I always go through and write notes as we approach release rather than as PRS are merged, and I think I saw that only maintainers can edit the wiki now?

I wonder whether adding a `needs-release-note` label would reduce the chance of changes being missing from the notes, as has happened recently? The label description could clarify how it is different to in NumPy i.e. there is no requirement on the author to write it, and either way a maintainer needs to write into the wiki.

mdhaber commented:
Sure; you mean a label that is automatically appended but needs to be manually removed by a maintainer?
But if you're referring to 1.13.0, the main factor for any missing notes in `stats` was the early branching of the release notes from the Wiki.

lucascolley commented:
> you mean a label that is automatically appended but needs to be manually removed by a maintainer?

Not even - just one that maintainers can add manually after merging an enhancement/deprecation/breaking PR, to say 'someone should write a release note before the release manager puts everything together'.

> the main factor for any missing notes in stats

I was referring to when the `from scipy import {numpy_function}` expired deprecation was missing.

nimish commented:
> > I can do this if you've got any good examples
> 
> Just something to appear under '`scipy.stats` improvements' in the release notes, e.g. https://github.com/scipy/scipy/releases/tag/v1.13.0.
> 
> > I always go through and write notes as we approach release rather than as PRS are merged, and I think I saw that only maintainers can edit the wiki now?
> 
> I wonder whether adding a `needs-release-note` label would reduce the chance of changes being missing from the notes, as has happened recently? The label description could clarify how it is different to in NumPy i.e. there is no requirement on the author to write it, and either way a maintainer needs to write into the wiki.

* The [Irwin-Hall](https://en.wikipedia.org/wiki/Irwin%E2%80%93Hall_distribution) or [Uniform Sum](https://mathworld.wolfram.com/UniformSumDistribution.html) has been added as `scipy.stats.irwinhall`

mdhaber commented:
Either way is fine. For me, one that is automatically added would be particularly useful. Before release, I check *all* PRs I've been involved in over the course of the release. A mark like this would narrow it down to the ones that still need attention.

lucascolley commented:
> The [Irwin-Hall](https://en.wikipedia.org/wiki/Irwin%E2%80%93Hall_distribution) or [Uniform Sum](https://mathworld.wolfram.com/UniformSumDistribution.html) has been added as scipy.stats.irwinhall

added, thanks!

nimish commented:
> > The [Irwin-Hall](https://en.wikipedia.org/wiki/Irwin%E2%80%93Hall_distribution) or [Uniform Sum](https://mathworld.wolfram.com/UniformSumDistribution.html) has been added as scipy.stats.irwinhall
> 
> added, thanks!

D'oh it should read:

* The [Irwin-Hall](https://en.wikipedia.org/wiki/Irwin%E2%80%93Hall_distribution) or [Uniform Sum](https://mathworld.wolfram.com/UniformSumDistribution.html) distribution has been added as scipy.stats.irwinhall
